src/algebra/polynomial/big_operators.lean

```diff
@@ -37,18 +37,78 @@ namespace polynomial
+list.sum_le_foldr_max nat_degree (by simp) nat_degree_add_le _
+
+lemma nat_degree_multiset_sum_le (l : multiset (polynomial α)) :
+  nat_degree l.sum ≤ (l.map nat_degree).foldr max max_left_comm 0 :=
+quotient.induction_on l (by simpa using nat_degree_list_sum_le)
+
+lemma nat_degree_sum_le (f : ι → polynomial α) :
+  nat_degree (∑ i in s, f i) ≤ s.fold max 0 (nat_degree ∘ f) :=
+by simpa using nat_degree_multiset_sum_le (s.val.map f)
+
+lemma degree_list_sum_le (l : list (polynomial α)) :
+  degree l.sum ≤ (l.map nat_degree).maximum :=
+begin
+  by_cases h : l.sum = 0,
+  { simp [h] },
+  { rw degree_eq_nat_degree h,
+    suffices : (l.map nat_degree).maximum = ((l.map nat_degree).foldr max 0 : ℕ),
+    { rw this,
+      simpa [this] using nat_degree_list_sum_le l },
+    rw list.maximum_eq_coe_foldr_max_of_ne_nil,
+    { congr },
+    contrapose! h,
+    rw [list.map_eq_nil] at h,
+    simp [h] }
+end
+
+lemma nat_degree_list_prod_le (l : list (polynomial α)) :
+  nat_degree l.prod ≤ (l.map nat_degree).sum :=
+begin
+  induction l with hd tl IH,
+  { simp },
+  { simpa using nat_degree_mul_le.trans (add_le_add_left IH _) }
+end
+
+lemma coeff_list_prod_of_nat_degree_le (l : list (polynomial α)) (n : ℕ)
+  (hl : ∀ p ∈ l, nat_degree p ≤ n) :
+  coeff (list.prod l) (l.length * n) = (l.map (λ p, coeff p n)).prod :=
+begin
+  induction l with hd tl IH,
+  { simp },
+  { have hl' : ∀ (p ∈ tl), nat_degree p ≤ n := λ p hp, hl p (list.mem_cons_of_mem _ hp),
+    simp only [list.prod_cons, list.map, list.length],
+    rw [add_mul, one_mul, add_comm, ←IH hl', mul_comm tl.length],
+    have h : nat_degree tl.prod ≤ n * tl.length,
+    { refine (nat_degree_list_prod_le _).trans _,
+      rw [←tl.length_map nat_degree, mul_comm],
+      refine list.sum_le_of_forall_le _ _ _,
+      simpa using hl' },
+    have hdn : nat_degree hd ≤ n := hl _ (list.mem_cons_self _ _),
+    rcases hdn.eq_or_lt with rfl|hdn',
+    { cases h.eq_or_lt with h' h',
+      { rw [←h', coeff_mul_degree_add_degree, leading_coeff, leading_coeff] },
+      { rw [coeff_eq_zero_of_nat_degree_lt, coeff_eq_zero_of_nat_degree_lt h', mul_zero],
+        exact nat_degree_mul_le.trans_lt (add_lt_add_left h' _) } },
+    { rw [coeff_eq_zero_of_nat_degree_lt hdn', coeff_eq_zero_of_nat_degree_lt, zero_mul],
+      exact nat_degree_mul_le.trans_lt (add_lt_add_of_lt_of_le hdn' h) } }
+end
+
+end semiring
+
 section comm_semiring
 variables [comm_semiring R] (f : ι → polynomial R) (t : multiset (polynomial R))
 
 lemma nat_degree_multiset_prod_le :
-  t.prod.nat_degree ≤ (t.map (λ f, nat_degree f)).sum :=
-begin
-  refine multiset.induction_on t _ (λ a t ih, _), { simp },
-  rw [prod_cons, map_cons, sum_cons],
-  transitivity a.nat_degree + t.prod.nat_degree,
-  { apply polynomial.nat_degree_mul_le },
-  { exact add_le_add (le_refl _) ih }
-end
+  t.prod.nat_degree ≤ (t.map nat_degree).sum :=
+quotient.induction_on t (by simpa using nat_degree_list_prod_le)
```

Comment:
Since some of the `multiset`/`finset` results work for `semiring` too, have you tried putting any of these in the `semiring` section to see if you can generalize it for free?